source/extensions/filters/http/geoip/geoip_filter.cc

```diff
@@ -71,49 +69,31 @@ Http::FilterHeadersStatus GeoipFilter::decodeHeaders(Http::RequestHeaderMap& hea
-  lookupGeolocationAnonData(config_->geoAnonHostingHeader(), remote_address,
-                            &Driver::getIsAnonymousHostingProvider);
-  lookupGeolocationAnonData(config_->geoAnonTorHeader(), remote_address,
-                            &Driver::getIsAnonymousTorExitNode);
-  lookupGeolocationAnonData(config_->geoAnonProxyHeader(), remote_address,
-                            &Driver::getIsAnonymousPublicProxy);
-  all_lookups_invoked_ = true;
-
-  // Stop the iteration for headers for the current filter and the filters following.
-  return allLookupsComplete() ? Http::FilterHeadersStatus::Continue
-                              : Http::FilterHeadersStatus::StopIteration;
-}
 
-void GeoipFilter::lookupGeolocationData(
-    const absl::optional<std::string>& geo_header,
-    const Network::Address::InstanceConstSharedPtr& remote_address,
-    void (Driver::*func)(const Network::Address::InstanceConstSharedPtr& address,
-                         const LookupCallbacks& callbacks,
-                         const absl::optional<std::string>& geo_header) const) {
-  if (geo_header) {
-    ++num_lookups_to_perform_;
-    ((*driver_).*func)(remote_address, *this, geo_header);
-  }
-}
+  ASSERT(driver_, "No driver is available to perform geolocation lookup");
+
+  // Capturing weak_ptr to GeoipFilter so that filter can be safely accessed in the posted callback.
+  // This is a safe measure to protect against the case when filter gets deleted before the callback
+  // is run.
+  GeoipFilterWeakPtr self = weak_from_this();
+  state_ = State::InProgress;
+  // Copy header values to pass to the driver lookup function (in case filter gets destroyed before
+  // lookup completes).
+  absl::flat_hash_set<std::string> geo_headers = config_->geoHeaders();
+  absl::flat_hash_set<std::string> geo_anon_headers = config_->geoAnonHeaders();
+  driver_->lookup(
+      LookupRequest{std::move(remote_address), std::move(geo_headers), std::move(geo_anon_headers)},
+      [self, &dispatcher = decoder_callbacks_->dispatcher()](LookupResult&& result) {
+        dispatcher.post([self, result]() {
+          if (GeoipFilterSharedPtr filter = self.lock()) {
+            filter->onLookupComplete(std::move(result));
+          }
+        });
+      });
 
-void GeoipFilter::lookupGeolocationAnonData(
-    const absl::optional<std::string>& geo_header,
-    const Network::Address::InstanceConstSharedPtr& remote_address,
-    void (Driver::*func)(const Network::Address::InstanceConstSharedPtr& address,
-                         const LookupCallbacks& callbacks,
-                         const absl::optional<std::string>& geo_header) const) {
-  if (geo_header) {
-    ++num_lookups_to_perform_;
-    ((*driver_).*func)(remote_address, *this, geo_header);
-  }
+  // Stop the iteration for headers for the current filter and the filters following.
+  return state_ == State::Complete ? Http::FilterHeadersStatus::Continue
```

Comment:
I think there's no way state_ can be State::Complete here now that you're using the dispatcher, except in unit tests if the mock dispatcher is allowed to just call the callback immediately *and* you have the driver also call its callback immediately. Having a production code branch that's only relevant for tests is best avoided - I'd recommend capturing at least one of the callbacks in unit tests, and calling it after decodeHeaders. That way the code here can be always StopIteration, and the test better reflects actual behavior.

(Otherwise you also need a bunch of unnecessary complexity around *not* calling continueDecoding from the callback if you're still inside decodeHeaders, etc.)

Also, I think with this change `state_` is no longer necessary at all. This is the only place where it's read from, so all the writes to it become meaningless.